File: tests/common.fsharp.tests/JsonNode.fs
module JsonNodeTests

open Xunit
open common.fsharp
open Faqt

[<Fact>]
let ``asJsonObject with null returns a failure`` () =
    let result = JsonNode.asJsonObject null
    result.Should().BeFailure()

[<Fact>]
let ``asJsonObject with non JSON objects returns a failure`` () =
    let gen = Gen.JsonNode.nonJsonObject

    Check.fromGen gen (fun node ->
        let result = JsonNode.asJsonObject node
        result.Should().BeFailure())

[<Fact>]
let ``asJsonObject with JSON objects returns a success`` () =
    let gen = Gen.jsonObject

    Check.fromGen gen (fun node ->
        let result = JsonNode.asJsonObject node
        result.Should().BeSuccess())

[<Fact>]
let ``asJsonArray with null returns a failure`` () =
    let result = JsonNode.asJsonArray null
    result.Should().BeFailure

[<Fact>]
let ``asJsonArray with non JSON arrays returns a failure`` () =
    let gen = Gen.JsonNode.nonJsonArray

    Check.fromGen gen (fun node ->
        let result = JsonNode.asJsonArray node
        result.Should().BeFailure())

[<Fact>]
let ``asJsonArray with JSON arrays returns a success`` () =
    let gen = Gen.jsonArray

    Check.fromGen gen (fun node ->
        let result = JsonNode.asJsonArray node
        result.Should().BeSuccess())

[<Fact>]
let ``asJsonValue with null returns a failure`` () =
    let result = JsonNode.asJsonValue null
    result.Should().BeFailure()

[<Fact>]
let ``asJsonValue with non JSON values returns a failure`` () =
    let gen = Gen.JsonNode.nonJsonValue

    Check.fromGen gen (fun node ->
        let result = JsonNode.asJsonValue node
        result.Should().BeFailure())

[<Fact>]
let ``asJsonValue with JSON values returns a success`` () =
    let gen = Gen.jsonValue

    Check.fromGen gen (fun node ->
        let result = JsonNode.asJsonValue node
        result.Should().BeSuccess())

File: tests/common.tests/JsonNode.cs
﻿using CsCheck;
using System.Text.Json.Nodes;
using Xunit;

namespace common.tests;

public class JsonNodeTests
{
    [Fact]
    public void AsJsonObject_fails_if_the_node_is_null()
    {
        var node = (JsonNode?)null;
        var result = node.AsJsonObject();
        result.Should().BeError();
    }

    [Fact]
    public void AsJsonObject_fails_if_the_node_is_not_a_json_object()
    {
        var generator = JsonNodeGenerator.NonJsonObject;

        generator.Sample(node =>
        {
            var result = node.AsJsonObject();
            result.Should().BeError();
        });
    }

    [Fact]
    public void AsJsonObject_succeeds_if_the_node_is_a_json_object()
    {
        var generator = JsonNodeGenerator.JsonObject;

        generator.Sample(node =>
        {
            var result = node.AsJsonObject();
            result.Should().BeSuccess().Which.Should().BeAssignableTo<JsonObject>();
        });
    }

    [Fact]
    public void AsJsonArray_fails_if_the_node_is_null()
    {
        var node = (JsonNode?)null;
        var result = node.AsJsonArray();
        result.Should().BeError();
    }

    [Fact]
    public void AsJsonArray_fails_if_the_node_is_not_a_json_array()
    {
        var generator = JsonNodeGenerator.NonJsonArray;
        generator.Sample(node =>
        {
            var result = node.AsJsonArray();
            result.Should().BeError();
        });
    }

    [Fact]
    public void AsJsonArray_succeeds_if_the_node_is_a_json_array()
    {
        var generator = JsonNodeGenerator.JsonArray;
        generator.Sample(node =>
        {
            var result = node.AsJsonArray();
            result.Should().BeSuccess().Which.Should().BeAssignableTo<JsonArray>();
        });
    }

    [Fact]
    public void AsJsonValue_fails_if_the_node_is_null()
    {
        var node = (JsonNode?)null;
        var result = node.AsJsonValue();
        result.Should().BeError();
    }

    [Fact]
    public void AsJsonValue_fails_if_the_node_is_not_a_json_value()
    {
        var generator = JsonNodeGenerator.NonJsonValue;

        generator.Sample(node =>
        {
            var result = node.AsJsonValue();
            result.Should().BeError();
        });
    }

    [Fact]
    public void AsJsonValue_succeeds_if_the_node_is_a_json_value()
    {
        var generator = JsonNodeGenerator.JsonValue;

        generator.Sample(node =>
        {
            var result = node.AsJsonValue();
            result.Should().BeSuccess().Which.Should().BeAssignableTo<JsonValue>();
        });
    }
}